### PR TITLE
[RFR] PostgreSQL database

### PR DESCRIPTION
**Pull request by Kmaschta**
## TODO
- [x] PostgreSQL connection
- [x] CRUD & queries lib
- [x] Migrations lib
- [x] User model & API endpoint
- [x] Tests

### CODE REVIEW

**Comment by Kmaschta** [comment]
@jpetitcolas I installed `db-migrate` instead of `migrat` which is seems to be outdated lib.

File: api/migrations/20160119145322-user.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+var type;
+var seed;
+var async = require('async');
+
+/**
+  * We receive the dbmigrate dependency from dbmigrate initially.
+  * This enables us to not have to rely on NODE_PATH.
+  */
+exports.up = function(db, callback) {
+    async.series([
```
Comment (djhi) [inline api/migrations/20160119145322-user.js]: no series to execute here

File: api/migrations/20160119145322-user.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+        db.createTable.bind(db, 'users', {
+            id: { type: 'int', primaryKey: true, autoIncrement: true },
+            email: { type: 'string'},
+            password: { type: 'string'},
+        }),
+    ], callback);
+};
+
+exports.down = function(db, callback) {
+    async.series([
```
Comment (djhi) [inline api/migrations/20160119145322-user.js]: no series to execute here

File: api/src/lib/crud.js [hunk 3]
```diff
@@ -0,0 +1,128 @@
+import koa from 'koa';
+import coBody from 'co-body';
+import koaRoute from 'koa-route';
+
+export default (queries, restricted, excludedRouteMethod) => {
+    const app = koa();
+    let q;
```
Comment (djhi) [inline api/src/lib/crud.js]: `q` meaning isn't obvious. Can we rename `q` to `repositoy` and `queries` to `repositoryFactory` ? I know it's longer but I favor meaningful names.

File: api/src/lib/db/queries/batchDelete.js [hunk 4]
```diff
@@ -0,0 +1,21 @@
+export default (client, table, fields, idFieldName) => {
+    return function* batchInsert(ids) {
+        if (!Array.isArray(ids)) {
+            ids = [ids];
+        }
+
+        const idsQuery = ids.reduce((query, id, index) => {
+            const fieldName = idFieldName + index;
+            query.parameters[fieldName] = id;
+            query.sql.push(`${fieldName}`);
```
Comment (djhi) [inline api/src/lib/db/queries/batchDelete.js]: `query.sql.push(fieldName)` ?

File: api/src/lib/db/queries/batchInsert.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+        const values = [];
+        let parameters = {};
+
+        const columns = raw ? fields : fields.filter(field => {
+            return (!idAutoGenerated && field !== idFieldName) && typeof entities[0][field] !== 'undefined';
+        });
+
+        entities.forEach((entity, index) => {
+            const value = getValueSubQuery(entity, index + 1, null, raw);
+            parameters = Object.assign({}, parameters, value.parameters);
```
Comment (djhi) [inline api/src/lib/db/queries/batchInsert.js]:
```js
parameters = {...parameters, ...value.parameters};
```

File: api/src/lib/db/queries/batchUpsert.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+    return function* batchUpsert(entities) {
+        let error = null;
+
+        try {
+            // copy the table structure without the constraint
+            yield client.query_(`CREATE TEMPORARY TABLE ${tempTable} AS SELECT * FROM ${table} WHERE true = false;`);
+            yield tempBatchInsert(entities);
+
+            const setQuery = secondaryFields.map(field => {
```
Comment (djhi) [inline api/src/lib/db/queries/batchUpsert.js]:
```js
const setQuery = secondaryFields.map(f => `${f}=${tempTable}.${f}`);
```

File: api/src/lib/db/queries/field.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    getFieldNames: (fieldMap) => {
+        return Object
+            .keys(fieldMap)
+            .reduce(
+                (fields, table) => {
+                    if (table === 'composite') {
+                        return fields
+                            .concat(fieldMap[table].map(field => {
+                                return field.replace(/^.*AS\s+/, '');
+                            }));
```
Comment (djhi) [inline api/src/lib/db/queries/field.js]:
```js
return fields.concat(fieldMap[table].map(field =>field.replace(/^.*AS\s+/, '')));
```

File: api/src/lib/db/queries/field.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            .keys(fieldMap)
+            .reduce(
+                (fields, table) => {
+                    if (table === 'composite') {
+                        return fields.concat(fieldMap[table]);
+                    }
+                    return fields
+                        .concat(fieldMap[table].map(field => {
+                            return `${table}.${field} AS ${field}`;
+                        }));
```
Comment (djhi) [inline api/src/lib/db/queries/field.js]: same

File: api/src/lib/db/queries/selectPage.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+        const whereParts = [];
+
+        // handle filter param
+        if (match && searchableFields.length > 0) {
+            const whereFilerParts = [];
+            match = '%' + match + '%';
+            params.match = match;
+            searchableFields.forEach(field => {
+                whereFilerParts.push(`${field}::text ILIKE $match`);
+            });
```
Comment (djhi) [inline api/src/lib/db/queries/selectPage.js]:
```js
searchableFields.forEach(field => whereFilerParts.push(`${field}::text ILIKE $match`));
```

File: api/src/lib/db/queries/selectPage.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+        }
+
+        // always sort by id to avoid randomness in case of identical sortField value
+        const sortQuery = [`${idFieldName} ASC`];
+        if (sort) {
+            sort = sort.toLowerCase();
+            if (specificSorts && specificSorts.hasOwnProperty(sort)) {
+                const specificSort = specificSorts[sort].reduce(
+                    (result, condition, index) => `${result} WHEN '${condition}' THEN ${index + 1}`,
+                    `CASE ${sort}`
```
Comment (djhi) [inline api/src/lib/db/queries/selectPage.js]: Is this a multiline template ? I'm not sure it's correct

**Comment by djhi** [comment]
You did a great job !

File: api/migrations/20160119145322-user.js [hunk 11]
```diff
@@ -0,0 +1,26 @@
+'use strict';
+
+var dbm;
```
Comment (jpetitcolas) [inline api/migrations/20160119145322-user.js]: Not used

File: api/src/lib/crud.js [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+    app.use(function* (next) {
+        q = queries(this.client);
+        yield next;
+    });
+
+    if (!excludedRouteMethod || excludedRouteMethod.indexOf('get') === -1) {
```
Comment (jpetitcolas) [inline api/src/lib/crud.js]: I would remove this logical of excluded routes. Instead, we are using a middleware checking the method and throwing a 404 if it is not allowed. For instance, something like (not tested code):

```
app.use(function* (next) {
        if (this.method !== 'GET') {
            this.status = 404;
            return;
        }

        yield next;
    });
```

Which is more readable and more flexible.

File: api/migrations/20160119145322-user.js [hunk 13]
```diff
@@ -0,0 +1,26 @@
...
+var type;
+var seed;
+var async = require('async');
```
Comment (jpetitcolas) [inline api/migrations/20160119145322-user.js]: Not sure we need this dependency as we are supposed to have Koa. Can't we deal with some `co` calls?

File: api/src/lib/crud.js (hunk 3 above)
Comment (Kmaschta) [inline api/src/lib/crud.js]: The lib is called `queries`, which represent a simple list of SQL queries.
I renamed `queries` to `queriesFactory` and `q` to `queries`.
Let me know if you're ok with that !

File: api/src/lib/db/queries/batchDelete.js (hunk 4 above)
Comment (Kmaschta) [inline api/src/lib/db/queries/batchDelete.js]: Oops ! There is a missing character. Good catch.

File: api/src/lib/db/queries/selectPage.js (hunk 10 above)
Comment (Kmaschta) [inline api/src/lib/db/queries/selectPage.js]: This is te original code, did I miss something ?
```js
if (specificSorts && specificSorts.hasOwnProperty(sort)) {
    var specificSort = specificSorts[sort].reduce(function (result, condition, index) {
        return result + ' WHEN \'' + condition + '\' THEN ' + (index + 1);
    }, 'CASE ' + sort);
    sortQuery.unshift(specificSort + ' END ' + (sortDir.toLowerCase() === 'asc' ? ' ASC' : ' DESC'));
} else if (sortableFields.indexOf(sort) !== -1) {
    sortQuery.unshift(sort + (sortDir.toLowerCase() === 'asc' ? ' ASC' : ' DESC'));
}
```

File: api/migrations/20160119145322-user.js (hunk 11 above)
Comment (Kmaschta) [inline api/migrations/20160119145322-user.js]: It is auto-generated by `db-migrate`, but you're right :+1:

File: api/migrations/20160119145322-user.js (hunk 13 above)
Comment (Kmaschta) [inline api/migrations/20160119145322-user.js]: Sure, due to precedent Gildas' comment, we don't need it.

File: api/src/lib/crud.js (hunk 12 above)
Comment (Kmaschta) [inline api/src/lib/crud.js]: I cleaned this file and append this middleware
```js
app.use(function* (next) {
    if (availableMethods.indexOf(this.method) === -1) {
        this.status = 405; // HTTP 405 Method Not Allowed
        this.body = {};
    }

    yield next;
});
```

Let me know if it suits you.

File: api/src/lib/db/queries/selectPage.js (hunk 10 above)
Comment (djhi) [inline api/src/lib/db/queries/selectPage.js]: nvm, missed the comma at L81...

File: api/src/lib/crud.js (hunk 3 above)
Comment (djhi) [inline api/src/lib/crud.js]: :+1:

File: api/migrations/20160119145322-user.js [hunk 14]
```diff
@@ -0,0 +1,19 @@
+'use strict';
+
+/**
+  * We receive the dbmigrate dependency from dbmigrate initially.
+  * This enables us to not have to rely on NODE_PATH.
+  */
+exports.up = function(db, callback) {
+    db.createTable.bind(db, 'users', {
```
Comment (djhi) [inline api/migrations/20160119145322-user.js]: you bind it, but you don't call the bound function

File: api/migrations/20160119145322-user.js [hunk 15]
```diff
@@ -0,0 +1,19 @@
...
+        id: { type: 'int', primaryKey: true, autoIncrement: true },
+        email: { type: 'string'},
+        password: { type: 'string'},
+    });
+    callback();
```
Comment (djhi) [inline api/migrations/20160119145322-user.js]:
```js
db.createTable(args, callback)
```

File: api/migrations/20160119145322-user.js [hunk 16]
```diff
@@ -0,0 +1,19 @@
...
+};
+
+exports.down = function(db, callback) {
+    db.dropTable.bind(db, 'users');
+    callback();
```
Comment (djhi) [inline api/migrations/20160119145322-user.js]: same

File: api/src/lib/crud.js [hunk 17]
```diff
@@ -0,0 +1,106 @@
+import koa from 'koa';
+import coBody from 'co-body';
+import koaRoute from 'koa-route';
+
+export default (queriesFactory, availableMethods = ['GET', 'POST', 'PUT', 'DELETE']) => {
+    const app = koa();
+    let queries;
```
Comment (djhi) [inline api/src/lib/crud.js]: much better !

File: api/src/lib/db/client.js [hunk 18]
```diff
@@ -0,0 +1,25 @@
+import pg from 'pg';
+import coPg from 'co-pg';
+import named from 'node-postgres-named';
+
+export default function* dbClient({host, port, user, password, database}) {
+    const postgresql = coPg(pg);
+    const connection = `postgres://${user}:${password}@${host}:${port}/${database}`;
+    const connectionResponse = yield postgresql.connect_(connection);
+    const client = connectionResponse[0];
+    const done = connectionResponse[1];
```
Comment (djhi) [inline api/src/lib/db/client.js]:
```js
const [client, done] = connectionResponse;
```

Comment (Kmaschta) [inline api/src/lib/db/client.js]: I always have this error, I tested many fixes, babel transformer, runtime, polyfill ... But I failed everytime.

```/home/kmaschta/Projects/Marmelab/javascript-boilerplate/api/src/lib/db/client.js:30
var _marked = [dbClient].map(_regeneratorRuntime.mark);
                             ^

ReferenceError: _regeneratorRuntime is not defined
    at Object.<anonymous> (client.js:5:34)
    at Module._compile (module.js:435:26)
    at loader (/home/kmaschta/Projects/Marmelab/javascript-boilerplate/node_modules/babel-register/lib/node.js:130:5)
    at Object.require.extensions.(anonymous function) [as .js] (/home/kmaschta/Projects/Marmelab/javascript-boilerplate/node_modules/babel-register/lib/node.js:140:7)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:313:12)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at Object.<anonymous> (/home/kmaschta/Projects/Marmelab/javascript-boilerplate/api/src/server.js:43:15)
    at Module._compile (module.js:435:26```

**Comment by Kmaschta** [comment]
Rebased

**Pull request merged**